add tests for examples/pastebin

ttdonovan opened the pull request:
Adding tests for examples/pastebin #62 

Any feedback is welcomed.

SergioBenitez commented:
I haven't had a chance to look over this yet, but I want to say thank you for taking this on! It's _very much_ so appreciated. :D

SergioBenitez commented on examples/pastebin/.gitignore:
```diff
@@ -0,0 +1 @@
+/upload/
```
I want to keep this folder around so that user's don't have to create it manually.

SergioBenitez commented on examples/pastebin/src/main.rs:
```diff
@@ -4,6 +4,8 @@
 extern crate rand;
 
+#[cfg(test)] mod tests;
+
```
We can kill this empty line.

SergioBenitez commented on examples/pastebin/src/tests/mod.rs:
```diff
@@ -0,0 +1,69 @@
+fn upload() {
+    let rocket = rocket::ignite().mount("/", routes![super::upload]);
+    let mut req = MockRequest::new(Method::Post, "/")
+        .body(include_str!("file.txt").to_string());
```
You shouldn't need to call `to_string` here.

SergioBenitez commented on examples/pastebin/src/tests/mod.rs:
```diff
@@ -0,0 +1,69 @@
+    let rocket = rocket::ignite().mount("/", routes![super::upload]);
+    let mut req = MockRequest::new(Method::Post, "/")
+        .body(include_str!("file.txt").to_string());
+    let mut response = req.dispatch_with(&rocket);
```
Add an empty line above and kill the empty one below.

SergioBenitez commented on examples/pastebin/src/tests/mod.rs:
```diff
@@ -0,0 +1,69 @@
+            match url.trim().split("/").last() {
+                Some(paste_id) => {
+                    let filename = format!("upload/{}", paste_id);
+                    assert!(Path::new(&filename).exists());
```
This is assuming a particular implementation, which I don't think these tests should do. I think instead we should combine this with the next test and do a `POST` followed by a `GET`. Would be cool to do several of them, in fact, to make sure everything looks good.

SergioBenitez commented on examples/pastebin/src/tests/mod.rs:
```diff
@@ -0,0 +1,69 @@
+}
+
+#[test]
+fn retrieve_file_found() {
```
This is interesting because it's peering into the implementation. I think it'd be better to do a `POST` and then a `GET` instead, as mentioned above. What if we decide not to use files in the future, for instance? Or if the file structure changes?

SergioBenitez commented on examples/pastebin/src/tests/mod.rs:
```diff
@@ -0,0 +1,69 @@
+#[test]
+fn retrieve_file_not_found() {
+    let rocket = rocket::ignite().mount("/", routes![super::retrieve]);
+    let mut req = MockRequest::new(Method::Get, format!("/{}", "file-not-found"));
```
Why use `format` here?

ttdonovan commented:
@SergioBenitez I am out of town until next week. I briefly read over some
of your comments but will most likely not be able to dig into them anytime
soon. Feel free to alter or scrap my pull request however you feel fit.
Rust is fairly new language to me and coming from Ruby there is still a lot
I need to learn. My one observation or commit about testing the
implementation is - I'm not certain I like the idea of combining the POST
with the GET. Wouldn't there be double testing going on with also testing
the POST? Sometimes I find it hard to draw the line between integration and
unit tests. This example I feel like some of the internal implementation
knowledge is required. I would be interested in seeing an approach from
another Rust developer.

On Thu, Jan 26, 2017 at 21:52 Sergio Benitez <notifications@github.com>
wrote:

> *@SergioBenitez* requested changes on this pull request.
>
>
>
>
> Overall, an awesome job! My main concern is that the tests are much too
> knowledgeable about the specific implementation. I think the tests should
> only know as much about the API as the API itself exposes.
> ------------------------------
>
>
>
>
> In examples/pastebin/.gitignore
> <https://github.com/SergioBenitez/Rocket/pull/115#pullrequestreview-18789089>
> :
>
>
> > @@ -0,0 +1 @@
>
> +/upload/
>
>
>
> I want to keep this folder around so that user's don't have to create it
> manually.
>
>
>
>
> ------------------------------
>
>
>
>
> In examples/pastebin/src/main.rs
> <https://github.com/SergioBenitez/Rocket/pull/115#pullrequestreview-18789089>
> :
>
>
> > @@ -4,6 +4,8 @@
>
>  extern crate rocket;
>
>  extern crate rand;
>
>
>
> +#[cfg(test)] mod tests;
>
> +
>
>
>
> We can kill this empty line.
>
>
>
>
> ------------------------------
>
>
>
>
> In examples/pastebin/src/tests/mod.rs
> <https://github.com/SergioBenitez/Rocket/pull/115#pullrequestreview-18789089>
> :
>
>
> > +fn index() {
>
> +    let rocket = rocket::ignite().mount("/", routes![super::index]);
>
> +    let mut req = MockRequest::new(Method::Get, "/");
>
> +    let mut response = req.dispatch_with(&rocket);
>
> +
>
> +    assert_eq!(response.status(), Status::Ok);
>
> +
>
> +    let body_str = response.body().and_then(|body| body.into_string());
>
> +    assert_eq!(body_str, Some(include_str!("usage.txt").to_string()));
>
> +}
>
> +
>
> +#[test]
>
> +fn upload() {
>
> +    let rocket = rocket::ignite().mount("/", routes![super::upload]);
>
> +    let mut req = MockRequest::new(Method::Post, "/")
>
> +        .body(include_str!("file.txt").to_string());
>
>
>
> You shouldn't need to call to_string here.
>
>
>
>
> ------------------------------
>
>
>
>
> In examples/pastebin/src/tests/mod.rs
> <https://github.com/SergioBenitez/Rocket/pull/115#pullrequestreview-18789089>
> :
>
>
> > +    let rocket = rocket::ignite().mount("/", routes![super::index]);
>
> +    let mut req = MockRequest::new(Method::Get, "/");
>
> +    let mut response = req.dispatch_with(&rocket);
>
> +
>
> +    assert_eq!(response.status(), Status::Ok);
>
> +
>
> +    let body_str = response.body().and_then(|body| body.into_string());
>
> +    assert_eq!(body_str, Some(include_str!("usage.txt").to_string()));
>
> +}
>
> +
>
> +#[test]
>
> +fn upload() {
>
> +    let rocket = rocket::ignite().mount("/", routes![super::upload]);
>
> +    let mut req = MockRequest::new(Method::Post, "/")
>
> +        .body(include_str!("file.txt").to_string());
>
> +    let mut response = req.dispatch_with(&rocket);
>
>
>
> Add an empty line above and kill the empty one below.
>
>
>
>
> ------------------------------
>
>
>
>
> In examples/pastebin/src/tests/mod.rs
> <https://github.com/SergioBenitez/Rocket/pull/115#pullrequestreview-18789089>
> :
>
>
> > +
>
> +#[test]
>
> +fn upload() {
>
> +    let rocket = rocket::ignite().mount("/", routes![super::upload]);
>
> +    let mut req = MockRequest::new(Method::Post, "/")
>
> +        .body(include_str!("file.txt").to_string());
>
> +    let mut response = req.dispatch_with(&rocket);
>
> +
>
> +    assert_eq!(response.status(), Status::Ok);
>
> +
>
> +    match response.body().and_then(|body| body.into_string()) {
>
> +        Some(url) => {
>
> +            match url.trim().split("/").last() {
>
> +                Some(paste_id) => {
>
> +                    let filename = format!("upload/{}", paste_id);
>
> +                    assert!(Path::new(&filename).exists());
>
>
>
> This is assuming a particular implementation, which I don't think these
> tests should do. I think instead we should combine this with the next test
> and do a POST followed by a GET. Would be cool to do several of them, in
> fact, to make sure everything looks good.
>
>
>
>
> ------------------------------
>
>
>
>
> In examples/pastebin/src/tests/mod.rs
> <https://github.com/SergioBenitez/Rocket/pull/115#pullrequestreview-18789089>
> :
>
>
> > +    match response.body().and_then(|body| body.into_string()) {
>
> +        Some(url) => {
>
> +            match url.trim().split("/").last() {
>
> +                Some(paste_id) => {
>
> +                    let filename = format!("upload/{}", paste_id);
>
> +                    assert!(Path::new(&filename).exists());
>
> +                },
>
> +                _ => unreachable!(),
>
> +            }
>
> +        },
>
> +        _ => unreachable!(),
>
> +    }
>
> +}
>
> +
>
> +#[test]
>
> +fn retrieve_file_found() {
>
>
>
> This is interesting because it's peering into the implementation. I think
> it'd be better to do a POST and then a GET instead, as mentioned above.
> What if we decide not to use files in the future, for instance? Or if the
> file structure changes?
>
>
>
>
> ------------------------------
>
>
>
>
> In examples/pastebin/src/tests/mod.rs
> <https://github.com/SergioBenitez/Rocket/pull/115#pullrequestreview-18789089>
> :
>
>
> > +    file.write_all(b"Hello from Rocket!").expect("failed to write file");
>
> +
>
> +    let rocket = rocket::ignite().mount("/", routes![super::retrieve]);
>
> +    let mut req = MockRequest::new(Method::Get, format!("/{}", id));
>
> +    let mut response = req.dispatch_with(&rocket);
>
> +
>
> +    assert_eq!(response.status(), Status::Ok);
>
> +
>
> +    let body_str = response.body().and_then(|body| body.into_string());
>
> +    assert_eq!(body_str, Some("Hello from Rocket!".to_string()));
>
> +}
>
> +
>
> +#[test]
>
> +fn retrieve_file_not_found() {
>
> +    let rocket = rocket::ignite().mount("/", routes![super::retrieve]);
>
> +    let mut req = MockRequest::new(Method::Get, format!("/{}", "file-not-found"));
>
>
>
> Why use format here?
>
>
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SergioBenitez/Rocket/pull/115#pullrequestreview-18789089>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAo3FQI_O6zSbVc6wbrDhDnM-w22zLJks5rWaIqgaJpZM4LdMXX>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

SergioBenitez commented:
Went ahead and implemented this in e6615af7e6aa1f71f7a6a69f4777a4530310272c. Thanks so much again for your contribution! :)
